packages/babel-helpers/src/helpers/applyDecs.js

```diff
@@ -14,11 +14,24 @@
   SETTER = 4;
 
   STATIC = 5;
+
+  CLASS = 10; // only used in assertValidReturnValue
 */
 
-function createMetadataMethodsForProperty(metadataMap, kind, property) {
+function createMetadataMethodsForProperty(
+  metadataMap,
+  kind,
+  property,
+  decorationState
+) {
   return {
     getMetadata: function (key) {
+      if (decorationState.finished) {
+        throw new Error(
```

Comment:
Nit: to minimize the helper size we can extract the error message into a template.